pos_voucher/models/pos_voucher.py

```diff
@@ -0,0 +1,151 @@
+                 'voucher_history_ids.consumed_date')
+    def _compute_date_amount(self):
+        for rec in self:
+            consumed_amount = sum(rec.voucher_history_ids.mapped('amount'))
+            rec.pending_amount = rec.total_amount - consumed_amount
+            if rec.voucher_history_ids:
+                consume_date = max(rec.voucher_history_ids.mapped('consumed_date'))
+                rec.consume_date = consume_date
+
+    @api.multi
+    def validate_voucher(self):
+        for rec in self:
+            vals = {'state': 'validated'}
+            if not rec.code or rec.code == '/':
+                code = rec.type_id.sequence_id.next_by_code(
+                    rec.type_id.sequence_id.code)
+                vals.update({'code': code})
+            rec.write(vals)
+        return True
+
+    @api.multi
+    def action_validate_vouchers(self):
+        states = list(set(self.mapped('state')))
+        if len(states) > 1 or states[0] != 'draft':
+            raise Warning('All selected records has to be in Draft State!')
+        return self.validate_voucher()
+
+    @api.multi
+    def consume_voucher(self, amount):
+        self.ensure_one()
+        if amount < self.pending_amount:
+            state = 'partially_consumed'
+        else:
+            state = 'consumed'
+        self.write({'state': state,
+                    'history_line_ids': [
+                        (0, 0, {'consumed_date': fields.Datetime.today(),
+                                'amount': amount,
+                                })]
+                    })
+        return True
+
+    @api.multi
+    def force_voucher_expired(self):
+        self.write({'state': 'expired', 'end_date': fields.Datetime.today()})
+        return True
+
+    @api.multi
+    def cancel_voucher(self):
+        self.write({'state': 'cancelled'})
+        return True
+
+    @api.model
+    def get_voucher_by_code(self, code, customer_id):
+        if not code or not customer_id:
+            return -1
+        voucher = self.env['pos.voucher'].search(
+            [('code', '=', code), ('customer_id', '=', customer_id)], limit=1)
+        if not voucher:
+            return -1, "Voucher code with the customer dosn't exist!"
+        else:
+            if voucher.state in ('validated', 'partially_consumed'):
+                return 1, voucher.read([])[0]
+            else:
+                return -1, 'Voucher is %s!' % str(voucher.state).title()
+
+
+class POSVoucherHistory(models.Model):
+    _name = 'pos.voucher.history'
+    _description = 'POS Voucher History'
+
+    pos_voucher_id = fields.Many2one('pos.voucher', 'Voucher', required=True)
+    consumed_date = fields.Datetime(required=True)
+    amount = fields.Float(required=True, digits=(16, 2))
+
+
+class POSOrder(models.Model):
+    _inherit = "pos.order"
+
+    voucher_id = fields.Many2one('pos.voucher', 'Voucher Used')
+
+    @api.model
+    def _order_fields(self, ui_order):
```

Comment:
This could be avoided if we process everything in `create_from_ui`.
`create_from_ui` could receive a kind of dict with voucher consumption details.

Also *keep in mind* that voucher consumption is not equal to order amount !!
Order could be paid by Cash + Voucher or Credit Card + Voucher or even Cash + Credit Card + Check + Voucher.

So.. 
- voucher consumed amount could be != voucher amount
- voucher consume amount could be != order amount